### ISSUE TITLE
/.cache/lvim/lvim.log: Too many open files

### ISSUE THREAD

**Issue by DucNgn** [LunarVim/LunarVim#3392]
### Problem description

1. Fresh Install of LunarVim (No errors during installation, installed all JS, Python, Rust dependencies)
2. Open `config.lua` to edit the file
3. Make some minimal changes. I was able to make the changes, such as uncomment, writing new lines, deleting, etc.
4. Save the file

Sometimes, it's able to save the file with no problem, but often it gives the error of `Too many open files`:

```
Error executing vim.schedule lua callback: ...uyen/.local/share/lunarvim/lvim/lua/lvim/config/init.lua:172: attempt to call field 'setup' (a nil value)
stack traceback:
        ...uyen/.local/share/lunarvim/lvim/lua/lvim/config/init.lua:172: in function <...uyen/.local/share/lunarvim/lvim/lua/lvim/config/init.lua:161>
Press ENTER or type command to continue
Error executing vim.schedule lua callback: ...uyen/.local/share/lunarvim/lvim/lua/lvim/config/init.lua:172: attempt to call field 'setup' (a nil value)
stack traceback:
        ...uyen/.local/share/lunarvim/lvim/lua/lvim/config/init.lua:172: in function <...uyen/.local/share/lunarvim/lvim/lua/lvim/config/init.lua:161>
Press ENTER or type command to continue
Error executing vim.schedule lua callback: ...packer/start/structlog.nvim/lua/structlog/sinks/file.lua:39: /Users/dnguyen/.cache/lvim/lvim.log: Too many
 open files
stack traceback:
        [C]: in function 'assert'
        ...packer/start/structlog.nvim/lua/structlog/sinks/file.lua:39: in function 'write'
        ...ack/packer/start/structlog.nvim/lua/structlog/logger.lua:58: in function 'log'
        ...ack/packer/start/structlog.nvim/lua/structlog/logger.lua:64: in function 'log'
        ...dnguyen/.local/share/lunarvim/lvim/lua/lvim/core/log.lua:151: in function 'add_entry'
        ...dnguyen/.local/share/lunarvim/lvim/lua/lvim/core/log.lua:205: in function 'warn'
        ...uyen/.local/share/lunarvim/lvim/lua/lvim/config/init.lua:129: in function 'load'
        ...uyen/.local/share/lunarvim/lvim/lua/lvim/config/init.lua:164: in function <...uyen/.local/share/lunarvim/lvim/lua/lvim/config/init.lua:161>
Press ENTER or type command to continue
Error executing vim.schedule lua callback: ...uyen/.local/share/lunarvim/lvim/lua/lvim/config/init.lua:172: attempt to call field 'setup' (a nil value)
stack traceback:
        ...uyen/.local/share/lunarvim/lvim/lua/lvim/config/init.lua:172: in function <...uyen/.local/share/lunarvim/lvim/lua/lvim/config/init.lua:161>
Press ENTER or type command to continue
Error executing vim.schedule lua callback: ...uyen/.local/share/lunarvim/lvim/lua/lvim/config/init.lua:172: attempt to call field 'setup' (a nil value)
stack traceback:
        ...uyen/.local/share/lunarvim/lvim/lua/lvim/config/init.lua:172: in function <...uyen/.local/share/lunarvim/lvim/lua/lvim/config/init.lua:161>
Press ENTER or type command to continue
Error executing vim.schedule lua callback: ...packer/start/structlog.nvim/lua/structlog/sinks/file.lua:39: /Users/dnguyen/.cache/lvim/lvim.log: Too many
 open files
stack traceback:
        [C]: in function 'assert'
        ...packer/start/structlog.nvim/lua/structlog/sinks/file.lua:39: in function 'write'
        ...ack/packer/start/structlog.nvim/lua/structlog/logger.lua:58: in function 'log'
        ...ack/packer/start/structlog.nvim/lua/structlog/logger.lua:64: in function 'log'
        ...dnguyen/.local/share/lunarvim/lvim/lua/lvim/core/log.lua:151: in function 'add_entry'
        ...dnguyen/.local/share/lunarvim/lvim/lua/lvim/core/log.lua:205: in function 'warn'
        ...uyen/.local/share/lunarvim/lvim/lua/lvim/config/init.lua:129: in function 'load'
        ...uyen/.local/share/lunarvim/lvim/lua/lvim/config/init.lua:164: in function <...uyen/.local/share/lunarvim/lvim/lua/lvim/config/init.lua:161>
Press ENTER or type command to continue
Error executing vim.schedule lua callback: ...packer/start/structlog.nvim/lua/structlog/sinks/file.lua:39: /Users/dnguyen/.cache/lvim/lvim.log: Too many
 open files
stack traceback:
        [C]: in function 'assert'
        ...packer/start/structlog.nvim/lua/structlog/sinks/file.lua:39: in function 'write'
        ...ack/packer/start/structlog.nvim/lua/structlog/logger.lua:58: in function 'log'
        ...ack/packer/start/structlog.nvim/lua/structlog/logger.lua:64: in function 'log'
        ...dnguyen/.local/share/lunarvim/lvim/lua/lvim/core/log.lua:151: in function 'add_entry'
        ...dnguyen/.local/share/lunarvim/lvim/lua/lvim/core/log.lua:205: in function 'warn'
        ...uyen/.local/share/lunarvim/lvim/lua/lvim/config/init.lua:129: in function 'load'
        ...uyen/.local/share/lunarvim/lvim/lua/lvim/config/init.lua:164: in function <...uyen/.local/share/lunarvim/lvim/lua/lvim/config/init.lua:161>
Press ENTER or type command to continue

```

### LunarVim version

master-375232c

### Neovim version (>= 0.8.0)

NVIM v0.8.0 Build type: Release LuaJIT 2.1.0-beta3 Compiled by brew@Monterey

### Operating system/version

macOS Ventura 13.0

### Steps to reproduce

1. Install lvim using the script on the website (the release version)
2. `lvim` with no file
3. `Leader + l + c` to edit the configuration.
4. Delete some commented lines of the default config.
5. Try saving with `Leader w`

Repeat step 4 and 5 until the error pops up.

### support info

`LspInfo`:
```
 Language client log: /Users/dnguyen/.local/state/nvim/lsp.log
Detected filetype:   lua

 1 client(s) attached to this buffer:

Client: sumneko_lua (id: 1, pid: nil, bufnr: [5])             
filetypes:       lua
autostart:       true
root directory:  Running in single file mode.
cmd:             lua-language-server

Configured servers list: sumneko_lua  
```

`LvimInfo`:
```
Buffer info
* filetype:                lua
* bufnr:                   5
* treesitter status:       active

Active client(s)
* name:                      sumneko_lua
* id:                        1
* supported filetype(s):     lua
* attached buffers:          5
* root_dir pattern:          5
* capabilities:              hoverProvider          | workspaceSymbolProvider | documentHighlightProvider
                             colorProvider          | typeDefinitionProvider  | documentFormattingProvider
                             referencesProvider     | definitionProvider      | documentRangeFormattingProvider
                             documentSymbolProvider | foldingRangeProvider

Automatic LSP info
* Skipped servers: [angularls, ansiblels, csharp_ls, cssmodules_ls, denols, ember, emmet_ls, eslint, glint, golangci_lint_ls, gradle_ls, graphql, jedi_language_server, ltex, phpactor, psalm, pylsp, quick_lint_js, reason_ls, rome, ruby_ls, solang, solc, sorbet, sourcery, sqlls, sqls, stylelint_lsp, svlangserver, tflint, verible, vuels]

Formatters info
* Active: 
* Supported: [lua_format, stylua]

Linters info
* Active: 
* Supported: [luacheck, selene]

Code actions info
* Active: 

```

### Screenshots

The error goes on for a while

<img width="1068" alt="image" src="https://user-images.githubusercontent.com/43587865/199299694-af4bb259-fcf1-436b-9418-e251cf266f51.png">

<img width="1065" alt="image" src="https://user-images.githubusercontent.com/43587865/199299814-64ad6d4f-644f-490d-ac5c-f453f6be6d4e.png">

If I keep pressing "enter" through it, sometimes, it shows a prompt:

`Ok to remove?` but with no name of the package.

If I keep pressing `N`. It shows the following:

<img width="1069" alt="image" src="https://user-images.githubusercontent.com/43587865/199301007-c86b3e0c-38f4-4814-9b54-f4f6b8a5e6fa.png">

It then corrupts my lvim, rendering it unusable until I reinstall everything.

**Comment by LostNeophyte** [comment]
#3391 is similar but different(I've talked with DucNgn on discord)

**Comment by LostNeophyte** [comment]
> Can confirm this is also happening to me. My config is not fresh though.

Just to confirm, you are not saving config.lua fast many times?

**Comment by LostNeophyte** [comment]
It looks like it only happens on mac, @minhtrannhat do you use a mac too?

**Comment by kobzar** [comment]
> > Can confirm this is also happening to me. My config is not fresh though.
> 
> Just to confirm, you are not saving config.lua fast many times?

This somethins other. 
Some time ago i fund that when i save config and run SPS + p + c
I see in logs line something like "you runned it twice"

**Comment by Vagab** [comment]
I(author of #3391) also use mac, haven’t tried it on other machines yet, but what I found is it also happens for me when I save it some number of times, but not necessarily swiftly. I have tried to increase the number of maximum open files(`launchctl limit maxfiles` was 256 unlimited) to `1000000 1000000`, but that didn’t help. So far what I was doing is accepting the inevitable and sometimes just reinstalling lvim and copying over the old configuration which is far from ideal

**Comment by LostNeophyte** [comment]
> but what I found is it also happens for me when I save it some number of times, but not necessarily swiftly. (`launchctl limit maxfiles` was 256 unlimited) to `1000000 1000000`, but that didn’t help. So far what I was doing is accepting the inevitable and sometimes just reinstalling lvim and copying over the old configuration which is far from ideal

what intervals between saving? (just a rough estimate)

> I have tried to increase the number of maximum open files

That was going to be my next suggestion. @Vagab it only happens in config.lua for you? 

To all: does `lvim.reload_config_on_save = false` work as a workaround?

**Comment by LostNeophyte** [comment]
when you save config.lua, an autocommand will run and run PackerSync and reload the config. the sync may take longer than 4-5 sec, but that depends on your machine and internet

**Comment by LostNeophyte** [comment]
> You all can try to put this line into your shell config file. That's what I use to suppress this noisy error.
> 
> ```shell
> ulimit -n 10240
> ```

Do you use a mac?

**Comment by LostNeophyte** [comment]
So it looks like it only happens on a mac (unless you save really fast)

**Comment by DucNgn** [comment]
I guess I'll stick with the workaround for now.

Just a quick thought but refreshing config on save as a default setting sounds expensive and somewhat unnecessary to me. Especially many of us have the habit of literally pressing save on any small changes.

**Comment by Vagab** [comment]
> To all: does lvim.reload_config_on_save = false work as a workaround?
yep, this works. Completely uncalled for suggestion for a solution that might not work: maybe it'd make sense to check if the reload is in progress and if it is, stop it and schedule a new one.
I would make a pr myself but I wouldn't even know where to start.

**Comment by LostNeophyte** [comment]
> > To all: does lvim.reload_config_on_save = false work as a workaround?
> 
> @LostNeophyte yep, this works. A completely uncalled for suggestion for solution that might not work: maybe it'd make sense to check if the reload is in progress and if it is, stop it and schedule a new one. I would make a pr myself but I wouldn't even know where to start.

Not sure if it's possible to stop, will look into the issue later (unless someone else wants to do so?) since the workaround works. I have a lot to do for this Friday and Monday but I'll probably find some time while procrastinating

**Comment by LostNeophyte** [comment]
#3431 will probably make it better, but not fix it

**Comment by Vagab** [comment]
@LostNeophyte won't sometimes then saving the config not reload it which is not what's expected by most users?

**Comment by LostNeophyte** [comment]
> @LostNeophyte won't sometimes then saving the config not reload it which is not what's expected by most users?

Before this PR the autocommand has been created every time on reload so you'd end up with 10 autocommands and the config would reload 10 times. After the PR it will be created once

**Comment by NurdinDev** [comment]
I have the same issue in macos my Lunarvim view is  release-1.2 and neovim 0.8

**Comment by LostNeophyte** [comment]
> I have the same issue in macos my Lunarvim view is release-1.2 and neovim 0.8

The fix is not included in 1.2, it's on master
